packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart

```diff
@@ -0,0 +1,227 @@
+import 'dart:convert';
+
+/// Detects keyboard layout according to the steps defined in
+/// `/usr/share/console-setup/pc105.tree`.
+class KeyboardLayoutDetector {
+  /// Creates a keyboard layout detector with the given steps.
+  KeyboardLayoutDetector(this._steps);
+
+  /// Create a keyboard layout detector from the given data.
+  factory KeyboardLayoutDetector.fromData(String data) {
+    final steps = _foldSteps(LineSplitter.split(data));
+    return KeyboardLayoutDetector(_parseSteps(steps));
+  }
+
+  final Map<int, KeyboardLayoutStep> _steps;
+
+  /// Returns the step for the given number or null if not found.
+  KeyboardLayoutStep? step(int number) => _steps[number];
+
+  // Folds a flat list of lines into a list of lines for each step.
+  static List<List<String>> _foldSteps(Iterable<String> lines) {
+    return lines.fold<List<List<String>>>([], (steps, line) {
+      if (line.startsWith('STEP ')) {
+        steps.add(<String>[line]);
+      } else {
+        assert(steps.isNotEmpty);
+        steps.last.add(line);
+      }
+      return steps;
+    });
+  }
+
+  // Parses steps from the given list of lines for each step.
+  static Map<int, KeyboardLayoutStep> _parseSteps(List<List<String>> lines) {
+    final steps = <int, KeyboardLayoutStep>{};
+    for (final step in lines) {
+      assert(lines.length >= 2);
+      final number = int.parse(step.removeAt(0).split(' ').last);
+      final type = step.first.split(' ').first;
+      switch (type) {
+        case 'PRESS':
+          steps[number] = PressKeyStep.parse(number, step);
+          break;
+        case 'FIND':
+        case 'FINDP':
+          steps[number] = FindKeyStep.parse(number, step);
+          break;
+        case 'MAP':
+          steps[number] = KeymapResultStep.parse(number, step);
+          break;
+        default:
+          throw UnimplementedError(type);
+      }
+    }
+    return steps;
+  }
+}
+
+/// The base class of keyboard layout detection steps.
+abstract class KeyboardLayoutStep {
+  /// Creates a keyboard layout detection step.
+  KeyboardLayoutStep(this.number);
+
+  /// The number of this step.
+  final int number;
+
+  /// Returns the next step.
+  int? next();
+}
+
+/// Ask the user to press a key on the keyboard.
+class PressKeyStep extends KeyboardLayoutStep {
+  /// Creates a key press step.
+  PressKeyStep(
+    int number, {
+    required this.keys,
+    required this.steps,
+  }) : super(number);
+
+  /// The list of keys that the user must press.
+  final List<String> keys;
+
+  /// The steps to take after the user has pressed the key.
+  final Map<int, int> steps;
+
+  /// The keycode user pressed;
+  late int _keycode;
+
+  // FIXME need to account for possible remapping. Find the API to translate
+  // kernel keycodes to X keycodes (xkb).
+  // MIN_KEYCODE = 8
```

Comment:
Is this last line of the comment supposed to be here? The following const seems to have replaced it.